dbus/dbus_connection_test.go

```diff
@@ -0,0 +1,89 @@
+// Copyright 2020 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+package dbus
+
+import (
+	"testing"
+	"unsafe"
+
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+
+	"github.com/mendersoftware/mender/dbus/mocks"
+	dmocks "github.com/mendersoftware/mender/dbus/mocks"
+)
+
+func TestNewConnectionLibGio(t *testing.T) {
+	api := &mocks.DBusAPI{}
+	conn, err := NewConnectionLibGio(api)
+	assert.NoError(t, err)
+	assert.NotNil(t, conn)
+}
+
+func TestConnectionOpen(t *testing.T) {
+	testCases := map[string]struct {
+		connNewErr error
+		outErr     error
+	}{
+		"ok": {
+			connNewErr: nil,
+			outErr:     nil,
+		},
+		"ko": {
+			connNewErr: errors.New("error"),
+			outErr:     errors.New("error"),
+		},
+	}
+
+	for name, tc := range testCases {
+		t.Run(name, func(t *testing.T) {
+			api := &dmocks.DBusAPI{}
+			defer api.AssertExpectations(t)
+
+			api.On("ConnectionNewForAddress",
+				mock.AnythingOfType("string"),
+				mock.AnythingOfType("int"),
+			).Return(unsafe.Pointer(nil), tc.connNewErr)
```

Comment:
It would be great, and fairly easy, to test this using a real DBus server. The key difference will be to connect to the user bus instead of the system bus (avoids root), and perhaps we could also switch interface prefix for testing purposes, to make sure we are not messing with anything real.

For connecting to the user bus, use the address in the `DBUS_SESSION_BUS_ADDRESS` environment variable. For connecting to the system bus, use `unix:path=/var/run/dbus/system_bus_socket` (defined in `/usr/share/dbus-1/system.conf`, but in practice I think it's universal).

Reply:
To be honest, for unit testing in the client and server packages, I really prefer using mocks, as it allows us to test edge cases (e.g. failure paths). Said that, I'll follow your suggestion and use a local dbus when testing it in action.